src/Evolution/DiscontinuousGalerkin/ProjectToBoundary.hpp

```diff
@@ -0,0 +1,190 @@
+}  // namespace detail
+
+template <typename VolumeVarsTagsList, typename FaceVarsTagsList, size_t Dim>
+void contiguously_project_to_boundary(
+    const gsl::not_null<Variables<FaceVarsTagsList>*> face_fields,
+    const Variables<VolumeVarsTagsList>& volume_fields,
+    const Mesh<Dim>& volume_mesh, const Direction<Dim>& direction) noexcept {
+  static_assert(tmpl::size<VolumeVarsTagsList>::value != 0,
+                "Must have non-zero number of volume fields");
+  static_assert(tmpl::size<FaceVarsTagsList>::value != 0,
+                "Must have non-zero number of face fields");
+  using face_vars_excluding_extras_at_end = tmpl::front<
+      tmpl::split_at<FaceVarsTagsList,
+                     tmpl::next<tmpl::index_of<
+                         FaceVarsTagsList, tmpl::back<VolumeVarsTagsList>>>>>;
+  using front_face_vars_split = tmpl::split_at<
+      face_vars_excluding_extras_at_end,
+      tmpl::index_of<FaceVarsTagsList, tmpl::front<VolumeVarsTagsList>>>;
+  using volume_vars_face_subset_list = tmpl::back<front_face_vars_split>;
+  static_assert(
+      std::is_same_v<volume_vars_face_subset_list, VolumeVarsTagsList>,
+      "The VolumeVarsTagsList must be a subset of the FaceVarsTagsList.");
+  constexpr const size_t number_of_independent_components =
+      Variables<VolumeVarsTagsList>::number_of_independent_components;
+  using first_volume_tag = tmpl::front<VolumeVarsTagsList>;
+
+  const Mesh<Dim> uniform_gauss_mesh(volume_mesh.extents(0),
+                                     volume_mesh.basis(0),
+                                     Spectral::Quadrature::Gauss);
+  if (volume_mesh == uniform_gauss_mesh) {
+    const Matrix identity{};
+    auto interpolation_matrices = make_array<Dim>(std::cref(identity));
+    const auto& matrix = Spectral::boundary_interpolation_matrices(
+        volume_mesh.slice_through(direction.dimension()));
+    gsl::at(interpolation_matrices, direction.dimension()) =
+        direction.side() == Side::Upper ? matrix.second : matrix.first;
+
+    auto& first_face_field = get<first_volume_tag>(*face_fields);
+    auto& first_volume_field = get<first_volume_tag>(volume_fields);
+
+    DataVector face_view{
+        first_face_field[0].data(),
+        first_face_field[0].size() * number_of_independent_components};
+
+    apply_matrices(make_not_null(&face_view), interpolation_matrices,
+                   // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+                   DataVector{const_cast<double*>(first_volume_field[0].data()),
+                              first_volume_field[0].size() *
+                                  number_of_independent_components},
+                   volume_mesh.extents());
+  } else {
+    ASSERT(Mesh<Dim>(volume_mesh.extents(0), volume_mesh.basis(0),
+                     Spectral::Quadrature::GaussLobatto) == volume_mesh,
+           "The mesh must be either a uniform Gauss or Gauss-Lobatto mesh, but "
+           "got "
+               << volume_mesh);
```

Comment:
- [non-binding] Mild suggestion for making a `uniform_gauss_lobatto_mesh` as you did before and using that in the assert
- Perhaps the message could be tweaked to clarify the operation would work in principle, but is just not implemented at this time?

Reply:
- I inlined it here because it's only needed in debug builds
- I agree, let me know if my wording is better